ownCloud/Client/Tasks/ScheduledTaskManager.swift

```diff
@@ -0,0 +1,243 @@
+				NotificationCenter.default.addObserver(self, selector: #selector(batteryLevelDidChange), name: UIDevice.batteryLevelDidChangeNotification, object: nil)
+			} else {
+				NotificationCenter.default.removeObserver(self, name: UIDevice.batteryLevelDidChangeNotification, object: nil)
+				lowBatteryDetected = false
+			}
+		}
+	}
+
+	private override init() {
+		super.init()
+		// Monitor app states
+		NotificationCenter.default.addObserver(self, selector: #selector(applicationStateChange), name: UIApplication.didEnterBackgroundNotification, object: nil)
+		NotificationCenter.default.addObserver(self, selector: #selector(applicationStateChange), name: UIApplication.didBecomeActiveNotification, object: nil)
+
+		// Monitor media upload settings changes
+		NotificationCenter.default.addObserver(self, selector: #selector(mediaUploadSettingsDidChange), name: UserDefaults.MediaUploadSettingsChangedNotification, object: nil)
+
+		// In iOS12 or later, activate Wifi monitoring
+		if #available(iOS 12, *) {
+			wifiMonitorQueue = DispatchQueue(label: "com.owncloud.scheduled_task_mgr.wifi_monitor")
+			wifiMonitor = NWPathMonitor(requiredInterfaceType: .wifi)
+			(wifiMonitor as? NWPathMonitor)?.pathUpdateHandler = { [weak self] path in
+				// Use "inexpensive" WiFi only (not behind a cellular hot-spot)
+				let wifiAvailable =  (path.status == .satisfied && !path.isExpensive)
+				self?.wifiDetected = wifiAvailable
+				if wifiAvailable {
+					self?.scheduleTasks()
+				}
+			}
+			(wifiMonitor as? NWPathMonitor)?.start(queue: wifiMonitorQueue!)
+		}
+
+		checkPowerState()
+
+		startMonitoringPhotoLibraryChangesIfNecessary()
+	}
+
+	deinit {
+		NotificationCenter.default.removeObserver(self)
+		if #available(iOS 12, *) {
+			(wifiMonitor as? NWPathMonitor)?.cancel()
+		}
+		stopMonitoringPhotoLibraryChanges()
+	}
+
+	// MARK: - Notifications handling
+
+	@objc private func applicationStateChange(notificaton:Notification) {
+		switch notificaton.name {
+		case UIApplication.didBecomeActiveNotification:
+			state = .foreground
+		case UIApplication.didEnterBackgroundNotification:
+			state = .background
+		default:
+			break
+		}
+
+		scheduleTasks()
+	}
+
+	@objc private func batteryLevelDidChange(notification:Notification) {
+		checkPowerState()
+		if lowBatteryDetected || externalPowerConnected {
+			scheduleTasks()
+		}
+	}
+
+	@objc private func mediaUploadSettingsDidChange(notification:Notification) {
+		if shallMonitorPhotoLibraryChanges() {
+			startMonitoringPhotoLibraryChangesIfNecessary()
+		} else {
+			stopMonitoringPhotoLibraryChanges()
+		}
+	}
+
+	// MARK: - Background fetching
+
+	func backgroundFetch(completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
+		self.state = .backgroundFetch
+		scheduleTasks(fetchCompletion: completionHandler)
+	}
+
+	// MARK: - Private methods
+
+	private func scheduleTasks(fetchCompletion:((UIBackgroundFetchResult) -> Void)? = nil, completion:((_ scheduledTaskCount:Int)->Void)? = nil) {
```

Comment:
There's currently no protection against:
- a task getting scheduled more than once at the same time (which definitely *can* make sense, but should be opt-in from the Task's side)
- the `state` changing *during* scheduling (could pose a problem for Tasks that rely on being called for every event they sign up for)

That led to a new photo being uploaded twice when I added an image to the Photo Library via the ownCloud app's "Open In" action. The photo upload task was apparently run twice in parallel.